sdk/identity/Azure.Identity/src/MsalClientBase.cs

```diff
@@ -61,7 +61,7 @@ protected async ValueTask<TClient> GetClientAsync(bool async, CancellationToken
 
             if (TokenCache != null)
             {
-                await TokenCache.RegisterCache(async, client.UserTokenCache, cancellationToken).ConfigureAwait(false);
+                await TokenCache.RegisterCache(async, client is IConfidentialClientApplication cca ? cca.AppTokenCache : client.UserTokenCache, cancellationToken).ConfigureAwait(false);
```

Comment:
I looked into this a bit more after we chatted about it. I believe in the case of `IConfidentialClientApplication` we need to actually regiser *both* the `AppTokenCache` *and* the  `UserTokenCache` because the latter might also be used in auth code authentication flows.